chore(docs): Update cStor user guide (#937)
openebs/openebs-docs: docs/ugcstor-csi.md

---
id: ugcstor-csi
title: cStor User Guide
sidebar_label: cStor
---
------
<br>

<img src="/docs/assets/svg/3-config-sequence.svg" alt="OpenEBS configuration flow" style="width:100%">

<br>
<br>

This user guide section provides the operations needed to be performed by the User and the Admin for configuring cStor related tasks. 

   :::note 
   The recommended approach to provision cStor Pools is to use cStorPoolCluster(CSPC), the detailed steps have been provided in this document. However, OpenEBS also supports provisioning of cStor Pools using StoragePoolClaim (SPC). For detailed instructions, refer to the <a href="https://docs.openebs.io/v260/docs/next/ugcstor.html" target="_blank">cStor User guide(SPC)</a>.<br>
   :::

## User operations

## Admin operations

[Expanding a cStor volume](#expanding-a-cstor-volume)

[Performance Tunings in cStor Pools](#performance-tunings-in-cstor-pools)

<h2><a class="anchor" aria-hidden="true" id="user-operations"></a>User Operations</h2>

<h2><a class="anchor" aria-hidden="true" id="user-operations"></a>Admin operations</h2>

### <a class="anchor" aria-hidden="true" id="expanding-a-cstor-volume"></a>Expanding a cStor volume

> OpenEBS cStor introduces support for expanding an iSCSI PV using the CSI provisioner. Provided cStor is configured to function as a CSI provisioner, you can expand iSCSI PVs that have been created by cStor CSI Driver. This feature is supported with Kubernetes versions 1.16 and above.

Reviewer: ```suggestion
OpenEBS cStor introduces support for expanding an iSCSI PV using the CSI provisioner. Provided cStor is configured to function as a CSI provisioner, you can expand iSCSI PVs that have been created by cStor CSI Driver. This feature is supported with Kubernetes versions 1.19 and above.
```
CSIDriver is only available in storage.k8s.io/v1 since k8s v1.19. @shubham14bajpai please check here once.

> OpenEBS cStor introduces support for expanding an iSCSI PV using the CSI provisioner. Provided cStor is configured to function as a CSI provisioner, you can expand iSCSI PVs that have been created by cStor CSI Driver. This feature is supported with Kubernetes versions 1.16 and above.

Author: Replaced the above text with,  "OpenEBS cStor introduces support for expanding a PersistentVolume using the CSI provisioner. Provided cStor is configured to function as a CSI provisioner, you can expand PVs that have been created by cStor CSI Driver."

<li>To resize a PV, edit the PVC definition and update the spec.resources.requests.storage to reflect the newly desired size, which must be greater than the original size.</li>
<li>The PV must be attached to a pod for it to be resized. There are two scenarios when resizing an cStor PV:<br>
 <ul>
    <li>If the PV is attached to a pod, cStor CSI driver expands the volume on the storage backend, re-scans the device and resizes the filesystem.</li>
    <li>When attempting to resize an unattached PV, cStor CSI driver expands the volume on the storage backend. Once the PVC is bound to a pod, the driver re-scans the device and resizes the filesystem. Kubernetes then updates the PVC size after the expansion operation has successfully completed.</li>
 </ul>
</ul>
<br>
Below example shows the way for expanding cStor volume and how it works. For an already existing StorageClass, you can edit the StorageClass to include the <code>allowVolumeExpansion: true</code>  parameter.

```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: cstor-sparse-auto
provisioner: cstor.csi.openebs.io

> allowVolumeExpansion: true

Author: Currently, for code blocks, we are using "Fenced code blocks" which do not take styling syntax. We will be trying out other ways.

busybox         1/1     Running   0          38m
```

To list PVCs, execute:

```
$ kubectl get pvc
```
Sample Output:
```
NAME                           STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS              AGE
cstor-pvc                      Bound    pvc-849bd646-6d3f-4a87-909e-2416d4e00904   5Gi        RWO            cstor-csi-disk            1d
```

To list PVs, execute:

```
$ kubectl get pv
```

Sample Output:

```
NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                   STORAGECLASS        REASON   AGE
pvc-849bd646-6d3f-4a87-909e-2416d4e00904   5Gi        RWO            Delete           Bound    default/cstor-pvc       cstor-csi-disk               40m
```

<br>
To resize the PV that has been created from 5Gi to 10Gi, edit the PVC definition and update the spec.resources.requests.storage to 10Gi. It may take a few seconds to update the actual size in the PVC resource, wait for the updated capacity to reflect in PVC status (pvc.status.capacity.storage). It is internally a two step process for volumes containing a file system:
<br>
<ul>
 <li>Volume expansion</li>

>  <li>FileSystem expansion</li>

Reviewer: ```suggestion
 <li>Filesystem expansion</li>
```

      poolConfig:
        dataRaidGroupType: mirror
        priorityClassName: utlra-priority
```
<font size="4">**Example configuration for Compression:**</font>
Compression values can be set at pool level only. There is no override mechanism like it was there in case of tolerations, resources, auxResources and priorityClass. Compression value must be one of on,off,lzjb,gzip,gzip-[1-9],zle and lz4.

**Note:** lz4 is the default compression algorithm that is used if the compression field is left unspecified on the cspc. Below is the sample yaml which has compression specified.

```

apiVersion: cstor.openebs.io/v1
kind: CStorPoolCluster
metadata:
  name: demo-pool-cluster
  namespace: openebs
spec:
  pools:
    - nodeSelector:
        kubernetes.io/hostname: worker-node-1

      dataRaidGroups:
      - cspiBlockDevices:
          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f36
          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f37

      poolConfig:
        dataRaidGroupType: mirror
        compression: lz
```
<font size="4">**Example configuration for Read Only Threshold:**</font>

> RO threshold can be set in a similar manner like compression. ROThresholdLimit is the threshold(percentage base) limit for pool read only mode. If ROThresholdLimit(%) amount of pool storage is consumed then the pool will be set to readonly. If ROThresholdLimit is set to 100 then entire pool storage will be used. By default it will be set to 85% i.e when unspecified on the CSPC.ROThresholdLimit value will be 0 < ROThresholdLimit <= 100. Following CSPC yaml has the ReadOnly Threshold percentage specified.

Reviewer 2: How about below one?
```suggestion
RO threshold can be set in a similar manner like compression. ROThresholdLimit is the threshold(percentage base) limit for pool read only mode. If storage consumption is greater than ROThresholdLimit(%) then the pool will be marked as readonly. If ROThresholdLimit is set to 100 then entire pool storage will be used. By default, it will be set to 85% i.e when unspecified on the CSPC.ROThresholdLimit value will be 0 < ROThresholdLimit <= 100. Following CSPC yaml has the ReadOnly Threshold percentage specified.
```

title: cStor User Guide
sidebar_label: cStor
---
------
<br>

<img src="/docs/assets/svg/3-config-sequence.svg" alt="OpenEBS configuration flow" style="width:100%">

<br>
<br>

This user guide section provides the operations needed to be performed by the User and the Admin for configuring cStor related tasks. 

   :::note 
   The recommended approach to provision cStor Pools is to use CStorPoolCluster(CSPC), the detailed steps have been provided in this document. However, OpenEBS also supports provisioning of cStor Pools using StoragePoolClaim (SPC). For detailed instructions, refer to the <a href="https://docs.openebs.io/v260/docs/next/ugcstor.html" target="_blank">cStor User guide(SPC)</a>.<br>
   :::

<h2> User operations</h2>

[Snapshot and Clone of a cStor Volume](#snapshot-and-clone-of-a-cstor-volume)

<h2> Admin operations</h2>

[Creating cStor storage pools](#creating-cstor-storage-pool)

[Creating cStor storage classes](#creating-cstor-storage-classes)

[Expanding a cStor volume](#expanding-a-cstor-volume)

[Performance Tunings in cStor Pools](#performance-tunings-in-cstor-pools)

## <a class="anchor" aria-hidden="true" id="user-operations"></a>User Operations

> ### <a class="anchor" aria-hidden="true" id="creating-cstor-storage-pool"></a>Snapshot and Clone of a cStor Volume

Reviewer 3: Just noticed , id is pointing to pool here instead of snapshot , can be fixed in upcoming PRs

Author: Added in the next PR.

---
id: ugcstor-csi
title: cStor User Guide
sidebar_label: cStor
---
------
<br>

<img src="/docs/assets/svg/3-config-sequence.svg" alt="OpenEBS configuration flow" style="width:100%">

<br>
<br>

This user guide section provides the operations needed to be performed by the User and the Admin for configuring cStor related tasks. 

   :::note 
   The recommended approach to provision cStor Pools is to use CStorPoolCluster(CSPC), the detailed steps have been provided in this document. However, OpenEBS also supports provisioning of cStor Pools using StoragePoolClaim (SPC). For detailed instructions, refer to the <a href="https://docs.openebs.io/v260/docs/next/ugcstor.html" target="_blank">cStor User guide(SPC)</a>.<br>
   :::

<h2> User operations</h2>

[Snapshot and Clone of a cStor Volume](#snapshot-and-clone-of-a-cstor-volume)

<h2> Admin operations</h2>

[Creating cStor storage pools](#creating-cstor-storage-pool)

[Creating cStor storage classes](#creating-cstor-storage-classes)

[Expanding a cStor volume](#expanding-a-cstor-volume)

> [Performance Tunings in cStor Pools](#performance-tunings-in-cstor-pools)

Reviewer 3: We can add volume tunnables as well using volume policy,  more info https://github.com/openebs/cstor-operators/blob/master/docs/tutorial/volumes/policy.md

Author: We will be adding these changes in the next PR.

```
kubectl get cspi -n openebs
```
Sample Output:

```
NAME                  HOSTNAME             ALLOCATED   FREE    CAPACITY   STATUS   AGE
cstor-disk-pool-vn92  worker-node-1        60k         9900M    9900M     ONLINE   2m17s
cstor-disk-pool-al65  worker-node-2        60k         9900M    9900M     ONLINE   2m17s
cstor-disk-pool-y7pn  worker-node-3        60k         9900M    9900M     ONLINE   2m17s
```
Once all the pods are in running state, these pool instances can be used for creation of cStor volumes.
<hr>

### <a class="anchor" aria-hidden="true" id="creating-cstor-storage-classes"></a>Creating cStor storage classes

StorageClass definition is an important task in the planning and execution of OpenEBS storage. The real power of CAS architecture is to give an independent or a dedicated storage engine like cStor for each workload, so that granular policies can be applied to that storage engine to tune the behaviour or performance as per the workload's need.
  
  #### Steps to create a cStor StorageClass:
   1. Decide the CStorPoolCluster for which you want to create a Storage Class.

>    2. Decide the replicaCount based on your requirement/workloads. OpenEBS doesn't restrict the replica count to set, but a <b>maximum of 5</b> replicas are allowed. It depends how users configure it, but for the availability of volumes <b>at least (n/2 + 1) replicas</b> should be up and connected to the target, where n is the replicaCount. The Replica Count should be greater than or equal to the number of cStor Pool Instances(CSPIs). The following are some example cases:

Reviewer 2: ```suggestion
   2. Decide the replicaCount based on your requirement/workloads. OpenEBS doesn't restrict the replica count to set, but a <b>maximum of 5</b> replicas are allowed. It depends how users configure it, but for the availability of volumes <b>at least (n/2 + 1) replicas</b> should be up and connected to the target, where n is the replicaCount. The Replica Count should be always less than or equal to the number of cStor Pool Instances(CSPIs). The following are some example cases:
```

    <ul> 
    <li>If a user configured replica count as 2, then always 2 replicas should be available to perform operations on volume.</li>
    <li>If a user configured replica count as 3 it should require at least 2 replicas should be available for volume to be operational.</li>
     <li>If a user configured replica count as 5 it should require at least 3 replicas should be available for volume to be operational.</li>
    </ul>
   3. Create a YAML spec file <code>cstor-csi-disk.yaml</code> using the template given below. Update the pool, replica count and other policies. By using this sample configuration YAML, a StorageClass will be created with 3 OpenEBS cStor replicas and will configure themselves on the pool instances. 

      ```
      kind: StorageClass
      apiVersion: storage.k8s.io/v1
      metadata:
        name: cstor-csi-disk
      provisioner: cstor.csi.openebs.io
      allowVolumeExpansion: true
      parameters:
        cas-type: cstor
        # cstorPoolCluster should have the name of the CSPC

>         cstorPoolCluster: cstor-storage

Reviewer 2: To be intact let's update the value to `cstor-disk-pool`(Since CSPC is created with the same name).

**Set priority class:** Sets the priority levels as required.

**Compression:** This helps in setting the compression for cStor pools.

**ReadOnly threshold:** Helps in specifying read only thresholds for cStor pools.
<br>

<font size="4">**Example configuration for Resource and Limits:**</font>

Following CSPC YAML specifies resources and auxResources that will get applied to all pool manager pods for the CSPC. Resources get applied to cstor-pool containers and auxResources gets applied to sidecar containers i.e. cstor-pool-mgmt and pool-exporter.
<br>
In the following CSPC YAML we have only one pool spec (@spec.pools). It is also possible to override the resource and limit value for a specific pool. 

```
apiVersion: cstor.openebs.io/v1
kind: CStorPoolCluster
metadata:
  name: cstor-disk-pool
  namespace: openebs
spec:
  resources:
    requests:
      memory: "2Gi"
      cpu: "250m"
    limits:
      memory: "4Gi"
      cpu: "500m"

  auxResources:
    requests:
      memory: "500Mi"
      cpu: "100m"
    limits:
      memory: "1Gi"
      cpu: "200m"
  pools:
    - nodeSelector:
        kubernetes.io/hostname: worker-node-1

      dataRaidGroups:

>       - cspiBlockDevices:

Reviewer: `cspiBlockDevices` is not in the schema that we follow... also with several other nitpicks, it should look something like this:
```yaml
apiVersion: cstor.openebs.io/v1
kind: CStorPoolCluster
metadata:
  name: cstor-disk-pool
  namespace: openebs
spec:
  resources:
    requests:
      memory: "64Mi"
      cpu: "250m"
    limits:
      memory: "128Mi"
      cpu: "500m"
  auxResources:
    requests:
      memory: "50Mi"
      cpu: "400m"
    limits:
      memory: "100Mi"
      cpu: "400m"
  
  pools:
    - nodeSelector:
        kubernetes.io/hostname: worker-node-1
      dataRaidGroups:
        - blockDevices:
            - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f36
            - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f37
      poolConfig:
        dataRaidGroupType: mirror
    
    - nodeSelector:
        kubernetes.io/hostname: worker-node-2
      dataRaidGroups:
        - blockDevices:
            - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f39
            - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f40
      poolConfig:
        dataRaidGroupType: mirror
    
    - nodeSelector:
        kubernetes.io/hostname: worker-node-3
      dataRaidGroups:
        - blockDevices:
            - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f42
            - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f43
      poolConfig:
        dataRaidGroupType: mirror
        resources:
          requests:
            memory: 70Mi
            cpu: 300m
          limits:
            memory: 130Mi
            cpu: 600m
        auxResources:
          requests:
            memory: 60Mi
            cpu: 500m
          limits:
            memory: 120Mi
            cpu: 500m
```

          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f37

      poolConfig:
        dataRaidGroupType: mirror
```

Following CSPC YAML explains how the resource and limits can be overridden. If you look at the CSPC YAML, there are no resources and auxResources specified at pool level for worker-node-1 and worker-node-2 but specified for worker-node-3. In this case, for worker-node-1 and worker-node-2 the resources and auxResources will be applied from @spec.resources and @spec.auxResources respectively but for worker-node-3 these will be applied from @spec.pools[2].poolConfig.resources and @spec.pools[2].poolConfig.auxResources respectively. 

```
apiVersion: cstor.openebs.io/v1
kind: CStorPoolCluster
metadata:
  name: cstor-disk-pool
  namespace: openebs
spec:
  resources:
    requests:
      memory: "64Mi"
      cpu: "250m"
    limits:
      memory: "128Mi"
      cpu: "500m"

  auxResources:
    requests:
      memory: "50Mi"
      cpu: "400m"
    limits:
      memory: "100Mi"
      cpu: "400m"

  pools:
    - nodeSelector:
        kubernetes.io/hostname: worker-node-1

      dataRaidGroups:

>       - cspiBlockDevices:

Reviewer: Same changes as the above comment here as well.

          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f43

      poolConfig:
        dataRaidGroupType: mirror
        resources:
          requests:
            memory: 70Mi
            cpu: 300m
          limits:
            memory: 130Mi
            cpu: 600m

        auxResources:
          requests:
            memory: 60Mi
            cpu: 500m
          limits:
            memory: 120Mi
            cpu: 500m

```

<font size="4">**Example configuration for Tolerations:**</font>

Tolerations are applied in a similar manner like resources and auxResources. The following is a sample CSPC YAML that has tolerations specified. For worker-node-1 and worker-node-2 tolerations are applied form @spec.tolerations but for worker-node-3 it is applied from @spec.pools[2].poolConfig.tolerations

```
apiVersion: cstor.openebs.io/v1
kind: CStorPoolCluster
metadata:
  name: cstor-disk-pool
  namespace: openebs
spec:

  tolerations:
  - key: data-plane-node
    operator: Equal
    value: true
    effect: NoSchedule

  pools:
    - nodeSelector:
        kubernetes.io/hostname: worker-node-1

      dataRaidGroups:

>       - cspiBlockDevices:

Reviewer: Same changes as above.

          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f36
          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f37

      poolConfig:
        dataRaidGroupType: mirror

    - nodeSelector:
        kubernetes.io/hostname: worker-node-2

      dataRaidGroups:
      - cspiBlockDevices:
          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f39
          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f40

      poolConfig:
        dataRaidGroupType: mirror

    - nodeSelector:
        kubernetes.io/hostname: worker-node-3

      dataRaidGroups:
      - cspiBlockDevices:
          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f42
          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f43

      poolConfig:
        dataRaidGroupType: mirror
        priorityClassName: utlra-priority
```
<font size="4">**Example configuration for Compression:**</font>

Compression values can be set at <b>pool level only</b>. There is no override mechanism like it was there in case of tolerations, resources, auxResources and priorityClass. Compression value must be one of 
- on
- off
- lzjb
- gzip 
- gzip-[1-9]
- zle

> - lz

Reviewer 3: this should be `lz4`

> **Note:** lz is the default compression algorithm that is used if the compression field is left unspecified on the cspc. Below is the sample yaml which has compression specified.

Reviewer 3: lz4 is the defaults  compression algorithm

```

apiVersion: cstor.openebs.io/v1
kind: CStorPoolCluster
metadata:
  name: demo-pool-cluster
  namespace: openebs
spec:
  pools:
    - nodeSelector:
        kubernetes.io/hostname: worker-node-1

      dataRaidGroups:
      - cspiBlockDevices:
          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f36
          - blockDeviceName: blockdevice-ada8ef910929513c1ad650c08fbe3f37

      poolConfig:
        dataRaidGroupType: mirror

>         compression: lz

Reviewer 3: change it to lz4 here as well
